masonry/src/passes/update.rs

```diff
@@ -402,6 +403,19 @@ pub(crate) fn run_update_focus_pass(root: &mut RenderRoot) {
     }
 
     let prev_focused = root.global_state.focused_widget;
+    let was_ime_active = root.global_state.is_ime_active;
+
+    let disable_ime = was_ime_active && prev_focused != root.global_state.next_focused_widget;
+    if disable_ime {
+        // IME was active, but the next focused widget is going to receive the Ime::Disabled event
+        // sent by the platform. Synthesize an Ime::Disabled event here and send it to the widget
+        // about to be unfocused.
+        run_on_text_event_pass(root, &TextEvent::Ime(winit::event::Ime::Disabled));
+        root.global_state.emit_signal(RenderRootSignal::EndIme);
```

Comment:
A much more targetted fix, with fewer non-local impacts might look like:
```suggestion
        // Disable the IME, which was enabled specifically for this widget.
        // Note that if the newly focused widget also requires IME, we will request it 
        // again - this resets the platform's state, ensuring that partial IME 
        // inputs do not "travel" between widgets
        root.global_state.emit_signal(RenderRootSignal::EndIme);
        if prev_focused == root.global_state.next_focused_widget {
             tracing::warn!(id = prev_focused.map(|id| id.trace()), "request_focus called whilst handling Ime::Disabled");
             // In this unlikely case, the rest of this handler will short-circuit, and IME will never be re-enabled 
             // for this widget. The resultant `ImeEnabled` event will be routed to this widget as it is the focused
             // widget. We don't handle this as above to avoid loops
            root.global_state.emit_signal(RenderRootSignal::StartIme);
        }
```

Reply:
I like it. We've hill-climbed the focus pass IME stuff to something no longer _as_ messy.

Reply:
I'm going to look into one more aspect of the focus update pass about the behavior surrounding disabling.